sdk/src/test/scala/com/horizen/consensus/HistoryConsensusCheckerTest.scala

```diff
@@ -0,0 +1,156 @@
+    println(s"append to History: ${newBlock.id}")
+    history.append(newBlock) match {
+      case Failure(ex) => {
+        println(s"Got exception during add block ${newBlock.id} to the history: ${ex.getMessage}")
+        throw ex
+      }
+      case Success(historyWithBlock) => historyWithBlock._1
+    }
+  }
+
+  private def historyUpdateShallBeFailed(history: SidechainHistory, newBlock: SidechainBlock, incorrectGenerationRules: GenerationRules): Unit = {
+    history.append(newBlock) match {
+      case Failure(ex) => {
+        println(s"Got expected exception during add block ${newBlock.id} to the history: ${ex.getMessage}")
+      }
+      case Success(historyWithBlock) => throw new IllegalStateException(s"Incorrect block generated by ${incorrectGenerationRules} had been successfully added")
+    }
+  }
+
+  private def generatorSelection(rnd: Random, generators: mutable.IndexedSeq[SidechainBlocksGenerator]): SidechainBlocksGenerator = {
+    @scala.annotation.tailrec
+    def generatorSelectionIteration(rnd: Random, index: Int, generators: mutable.IndexedSeq[SidechainBlocksGenerator]): SidechainBlocksGenerator = {
+      if (index == 0) {
+        generators(0)
+      }
+      else {
+        if (rnd.nextBoolean()) {
+          println()
+          println(s"return ${index} generator from ${generators.size - 1}")
+          generators(index)
+        }
+        else {
+          generatorSelectionIteration(rnd, index - 1, generators)
+        }
+      }
+    }
+    generatorSelectionIteration(rnd, generators.size - 1, generators)
+  }
+
+  def testWithSeed(testSeed: Int): Unit = {
+    //val testSeed = 234
+    val rnd: Random = new Random(testSeed)
+
+    val initialParams = TestNetParams(consensusSlotsInEpoch = 10, sidechainGenesisBlockTimestamp = 1333344452L)
+    val (params, genesisBlock, genesisGenerator, genesisForgingData, genesisEndEpochInfo) = SidechainBlocksGenerator.startSidechain(1000000L, testSeed, initialParams)
+    val history: SidechainHistory = createHistory(params, genesisBlock, genesisEndEpochInfo)
+    history.applyStakeConsensusEpochInfo(genesisBlock.id, genesisEndEpochInfo.stakeConsensusEpochInfo)
+    println(s"//////////////// Genesis epoch ${genesisBlock.id} had been ended ////////////////")
+
+    val historyUpdateAtEpochEnd = historyOperationOnEpochEnd(history) _
+
+    val generators = mutable.IndexedSeq(genesisGenerator)
+
+
+    (1 to 300)
+      .foldLeft[(SidechainHistory, mutable.IndexedSeq[SidechainBlocksGenerator])]((history, generators)) { (acc, index) =>
+        val currentHistory: SidechainHistory = acc._1
+        val currentGenerators: mutable.IndexedSeq[SidechainBlocksGenerator] =  acc._2
+
+        val nextGenerator: SidechainBlocksGenerator = generatorSelection(rnd, currentGenerators)
+        val nextCorrectGenerationRules: GenerationRules = GenerationRules.generateCorrectGenerationRules(rnd, nextGenerator)
```

Comment:
we should guarantee that the main scenarios are verified both for failures and success